test/server.ts

```diff
@@ -237,6 +244,135 @@ test('store server started in the background should use store location wanted by
+    if (options.onProcessStarted) {
+      (options.onProcessStarted)(serverProcess, serverProcessDeferred)
+    }
+    serverProcess.on('close', (code: number) => {
+      if (options.onProcessClosed) {
+        (options.onProcessClosed)(serverProcess)
+      }
+      for (let j = 0; j < serverProcessList.length; j++) {
+        if (serverProcessList[j][0] === serverProcess) {
+          serverProcessList.splice(j, 1)
+          break
+        }
+      }
+      if (code !== 0) {
+        serverProcessDeferred.reject(new Error(`A server process unexpectedly exited with code ${code}`))
+      } else {
+        serverProcessDeferred.resolve()
+      }
+    });
+    completedPromise = completedPromise.then(() => serverProcessDeferred.promise)
+  }
+  await completedPromise
+  if (timeoutObj !== null) {
+    clearTimeout(timeoutObj)
+  }
+}
+
+test('parallel server starts against the same store should result in a winner within 10 seconds', async (t: tape.Test) => {
+  await testParallelServerStart({
+    onTimeout: async (serverProcessList: ServerProcessList) => {
+      if (serverProcessList.length > 1) {
+        throw new Error('There are still multiple server processes running after 10 seconds')
+      } else if (serverProcessList.length > 0) {
+        t.ok(true, `only server ${serverProcessList[0][0].pid} is running after 10 seconds`)
+      } else {
+        // An issue out of scope of this test occured.
+      }
+    },
+    timeoutMillis: 10000,
+  })
+})
+
+test('parallel server starts against the same store should result in losers shutting down within 2 seconds after detecting they lost', async (t: tape.Test) => {
+  const serverProcessStates: {
+    [pid: number]: {
+      didWaitForProcessTermination: boolean,
+      terminationTimeout: NodeJS.Timer|null,
+    },
+  } = {}
+  const terminationTimeoutMillis = 2000
+  await testParallelServerStart({
+    onProcessClosed: (serverProcess: ChildProcess) => {
+      const state = serverProcessStates[serverProcess.pid]
+      delete serverProcess[serverProcess.pid]
+      if (state.terminationTimeout !== null) {
+        clearTimeout(state.terminationTimeout)
+        state.terminationTimeout = null
+      }
+    },
+    onProcessStarted: (serverProcess: ChildProcess, serverProcessDeferred: Deferred<void>) => {
+      serverProcessStates[serverProcess.pid] = {
+        didWaitForProcessTermination: false,
+        terminationTimeout: null,
+      }
+      const state = serverProcessStates[serverProcess.pid]
+      function updateExpectTermination (line: string) {
+        if (state.didWaitForProcessTermination) {
+          return
+        }
+        const lineNormalized = line.replace(/^\s+|\s+$/g, '').replace(/\s+/g, ' ').toLowerCase()
+        if (!lineNormalized.startsWith('canceling startup of server ')) {
+          return
+        }
+        console.log(`Server ${serverProcess.pid} detected another server started and should abort`)
```

Comment:
use t.pass here and specify a t.plan at the start of the test. In that case, if t.pass will be executed less than planned, the test will fail.